presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java

```diff
@@ -138,36 +142,38 @@ public void matches(PlanMatchPattern pattern)
 
         inTransaction(session -> {
             Map<PlanNodeId, PlanNodeCost> planNodeCosts = costCalculator.calculateCostForPlan(session, types, actual);
-            assertPlan(session, metadata, costCalculator, new Plan(actual, types, planNodeCosts), ruleApplication.lookup, pattern);
+            assertPlan(session, metadata, costCalculator, new Plan(actual, types, planNodeCosts), result.lookup, pattern);
             return null;
         });
     }
 
-    private RuleApplication applyRule()
+    private OptimizationResult optimize()
     {
-        SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
-        Memo memo = new Memo(idAllocator, plan);
-        Lookup lookup = Lookup.from(planNode -> Stream.of(memo.resolve(planNode)));
-
-        PlanNode memoRoot = memo.getNode(memo.getRootGroup());
-
-        return inTransaction(session -> applyRule(rule, memoRoot, ruleContext(symbolAllocator, lookup, session)));
-    }
-
-    private static <T> RuleApplication applyRule(Rule<T> rule, PlanNode planNode, Rule.Context context)
-    {
-        PlanNodeMatcher matcher = new PlanNodeMatcher(context.getLookup());
-        Match<T> match = matcher.match(rule.getPattern(), planNode);
-
-        Optional<PlanNode> result;
-        if (!rule.isEnabled(context.getSession()) || match.isEmpty()) {
-            result = Optional.empty();
-        }
-        else {
-            result = rule.apply(match.value(), match.captures(), context);
-        }
+        return inTransaction(session -> {
+            List<RuleApplicationListener.RuleApplication> ruleApplications = new ArrayList<>();
+            IterativeOptimizer optimizer = new IterativeOptimizer(
+                    new StatsRecorder(),
+                    ruleApplications::add,
+                    ImmutableSet.of(rule));
+
+            SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
+            Memo memo = new Memo(idAllocator, plan);
+            Lookup lookup = Lookup.from(planNode -> Stream.of(memo.resolve(planNode)));
+            optimizer.optimize(memo, session, symbolAllocator, idAllocator);
+
+            List<RuleApplicationListener.RuleApplication> nonEmptyRuleApplications = ruleApplications.stream()
+                    .filter(application -> application.getResult().isPresent())
+                    .collect(toImmutableList());
+
+            if (nonEmptyRuleApplications.size() == 1) {
+                return new OptimizationResult(lookup, symbolAllocator.getTypes(), Optional.of(memo.getNode(memo.getRootGroup())));
+            }
+            else if (nonEmptyRuleApplications.size() == 0) {
+                return new OptimizationResult(lookup, symbolAllocator.getTypes(), Optional.empty());
+            }
 
-        return new RuleApplication(context.getLookup(), context.getSymbolAllocator().getTypes(), result);
+            throw new IllegalStateException("Unexpected multiple rule executions");
```

Comment:
If rule applies several times, this is sufficient. For rule that applies indefinitely, this won't be sufficient though. Consider using `ruleListener` that matches once.